rust/src/io/reader.rs

```diff
@@ -335,7 +380,27 @@ async fn read_batch(
             row_id_arr,
         )?;
     }
-    Ok(batch)
+
+    let deletion_mask = match deletion_vector {
+        None | Some(DeletionVector::NoDeletions) => None,
+        Some(dvec) => Some(BooleanArray::from(
+            ids_in_batch_and_offset
+                .unwrap()
+                .0
+                .iter()
+                .map(|&id| match &dvec {
+                    DeletionVector::Bitmap(bitmap) => bitmap.contains(id as u32),
+                    DeletionVector::Set(hash_set) => hash_set.contains(&(id as u32)),
+                    _ => panic!("should never reach here"),
+                })
+                .collect::<Vec<_>>(),
+        )),
```

Comment:
I wonder if it would be more performant to move the match statement outside of the loop. Maybe even make this a method on the DeletionVector:

```rust
impl DeletionVector {
    fn build_predicate(&self, row_id_array: &UInt32Array) -> BooleanArray {
        match self {
            Self::Bitmap(bitmap) => ...
            Self::Set(set) => ...
        }
    }
}
```